### PR TITLE
Sparrow dom/confirmation polling

### PR DESCRIPTION
**Pull request by sparrowDom**
First pull request? Read our [guide to contributing](http://docs.originprotocol.com/#contributing)

### Checklist:

- [x] Code contains relevant tests for the problem you are solving
- [x] Ensure all new and existing tests pass
- ~[ ] Update any relevant READMEs and [docs](https://github.com/OriginProtocol/docs)~

**TODO:**
- in case we decide to merge this, tests still need to be written

### Description:
Metamask 4.12.0 introduced some breaking changes where `confirmation` and `receipt` events are not always triggered. And sometimes an error is thrown in spite of blockchain transaction succeeding. 

The solution this PR proposes: 
- for easier handling refactor so that contract function calls and contract deploys are handled by the same function
- keep `confirmation` and `receipt` events present and stop fallback solution in case those events are triggered
- when submitting transaction start a fallback pooling function that: 
  - continually pools for transaction receipt. If it finds it and `confirmation` event has not been triggered. Manually call confirmation callback
  - if transaction receipt has not been returned as promise resolution, mock it and resolve promise with it. 
- when we receive an error from Metamask that we know is not breaking don't reject the promise. Let fallback function handle resolution. 

**Related issues:**
- https://github.com/OriginProtocol/origin-dapp/issues/639
- https://github.com/OriginProtocol/origin-dapp/issues/673

**Pull request migrated from:**https://github.com/OriginProtocol/origin-js/pull/599

### CODE REVIEW

**Comment by sparrowDom** [comment]
@franckc a `Web3Mock` object is now added as a test helper that simulates different behaviours that can be observed with different versions of Metamask. 

Oh and maybe a quick shout to @tyleryasaka and @nick: if you ever (in the future) need to test the interaction between `contract-service` and `web3` this mock web3 object might be useful to you.

File: origin-js/test/helpers/web3-mock.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
+const EventEmitter = require('events')
+class MyEmitter extends EventEmitter {}
+
+//TODO document stuff here
+export default class Web3Mock {
+  constructor({
+    emitReceipt = true,
+    emitTransactionHash = true,
+    emitConfirmation = true,
+    nrOfConfirmations = 3,
```
[the remark this answers is not in the document]
Comment (sparrowDom) [reply origin-js/test/helpers/web3-mock.js]: 👍

File: origin-js/src/services/contract-service.js [hunk 2]
```diff
@@ -193,61 +195,133 @@ class ContractService {
...
 
     // transaction not mined
-    if (transactionInfo === null || transactionInfo.blockNumber === null){
+    if (transactionReceipt === null || transactionReceipt.blockNumber === null){
       setTimeout(() => {
-        this.checkForDeploymentCompletion(hash, confirmationCallback)
+        this.checkForTransactionCompletion(hash, contract, confirmationCallback, resolveCallback, promiseStatus)
       }, 1500)
     } else {
+      promiseStatus.numberOfTimesTransactionFoundByFallbackFcnt += 1
```
[the remark this answers is not in the document]
Comment (sparrowDom) [reply origin-js/src/services/contract-service.js]: :D this reminds me of a function name recovered from symbol files of the first Carmageddon game: https://github.com/jeff-1amstudios/carmageddon1-symbol-dump/blob/651ae28e6a097f6c49b7aa310004b3f0495f0e48/symbols/DETHRACE/source/common/pedestrn.c#L102

**Comment by sparrowDom** [comment]
@franckc the variables have been renamed. Deploying to staging first and testing it out sounds good.

**Comment by franckc** [comment]
Going to merge this PR in shortly then will deploy to staging this morning...

**Comment by sparrowDom** [comment]
nice!

**Pull request merged**